scripts/solr_builder/solr_builder/solr_builder.py

```diff
@@ -1,70 +1,599 @@
+        self._conn = None  # type: psycopg2._psycopg.connection
+        self.cache = dict()
+        self.cached_work_editions = set()
+        self.ia_cache = dict()
+        self.ia = None
+
+    def __enter__(self):
+        """
+        :rtype: LocalPostgresDataProvider
+        """
+        self._conn = psycopg2.connect(**self._db_conf)
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.clear_cache()
+        self._conn.close()
+
+    def query_all(self, query, cache_json=False):
+        """
+
+        :param str query:
+        :param bool cache_json:
+        :rtype: list
+        """
+        cur = self._conn.cursor()
+        cur.execute(query)
+        rows = cur.fetchall()
+        cur.close()
+
+        if rows:
+            if cache_json:
+                self.cache.update({row[0]: row[1] for row in rows})
+            return rows
+        else:
+            return []
+
+    def query_iter(self, query, size=20):
+        cur = self._conn.cursor()
+        cur.execute(query)
+
+        while True:
+            rows = cur.fetchmany(size)
+            if not rows:
+                break
+            for row in rows:
+                yield row
+
+        cur.close()
+
+    def query_batched(self, query, size, cursor_name=None, cache_json=False):
+        """
+        :param str query:
+        :param int size:
+        :param str or None cursor_name: if wanting to use a specific cursor
+        :param bool cache_json: Requires the select statement to be "Key", "JSON"
+        :return:
+        """
+        # Not sure if this name needs to be unique
+        cursor_name = cursor_name \
+            or 'solr_builder_server_side_cursor_' + uuid.uuid4().hex
+        cur = self._conn.cursor(name=cursor_name)
+        cur.itersize = size
+        cur.execute(query)
+
+        while True:
+            rows = cur.fetchmany(size)
+            if not rows:
+                break
+            else:
+                if cache_json:
+                    self.cache.update({row[0]: row[1] for row in rows})
+                yield rows
+
+        cur.close()
+
+    @staticmethod
+    async def _get_lite_metadata(ocaids, _recur_depth=0, _max_recur_depth=3):
+        """
+        For bulk fetch, some of the ocaids in Open Library may be bad
+        and break archive.org ES fetches. When this happens, we (up to
+        3 times) recursively split up the pool of ocaids to do as many
+        successful sub-bulk fetches as we can and then when limit is
+        reached, downstream code will fetch remaining ocaids individually
+        (and skip bad ocaids) 
+        """
+        if not ocaids or _recur_depth > _max_recur_depth:
+            logger.warning('Max recursion exceeded trying fetch IA data', extra={
+                'ocaids': ocaids
+            })
+            return []
+
+        r = requests.get("https://archive.org/advancedsearch.php", params={
```

Comment:
It would be nice if we DRY'd our calls to advancedsearch. We make calls to this API in several places throughout the code and it would be nice to be able to test it in/from 1 location